text/0058-valuecastable-format.md

```diff
@@ -0,0 +1,128 @@
+]
+```
+
+prints:
+
+```
+Value in binary: 00010010.00110100
+Value in hexadecimal: 12.34
+```
+
+However, sometimes it is also useful to print the raw value of such a signal. A shorthand is provided for that purpose:
+
+```py
+m.d.comb += Print(Format("Value: {num:x} (raw: {num!v:x})", num=num))
+```
+
+```
+Value: 12.34 (raw: 1234)
+```
+
+## Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A new overridable method is added to `ShapeCastable`:
+
+- `format(self, value: ValueCastable, format_desc: str) -> Format`
+
+When a `ValueCastable` is formatted without specifying a conversion (ie. `"!r"` or `"!v"`):
+
+- `shape()` is called
+- if the shape is a `ShapeCastable` and has a `format` method, the value-castable being formatted and the format descriptor (the part after `:` in `{}`, if any) are passed directly to `shape.format()`, and the result (which must be a `Format` object) is included directly in place of the format specifier
+- otherwise (the shape is a plain `Shape`, or doesn't have `format`), `Value.cast()` is called on the value-castable, and formatting proceeds as if it was a plain value
+
+A new conversion, `!v`, is added to `Format` (with syntax analogous to `!r`). When specified, the value being formatted is converted through `Value.cast` before proceeding with further formatting. It can be used to print the raw value of value-castables. It is a no-op when used with a plain `Value`.
+
+An implementation of `__format__` is added to the `ValueCastable` base class that always raises a `TypeError`, directing the user to `Format` instead (like the one that already exists on plain `Value`).
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+A new reserved name, `format`, is added to `ShapeCastable`, which is intended to be a fairly slim interface.
+
+The `__format__` on `ValueCastable` is the first time we have a method with an actual implementation.
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The `format` hook is added on `ShapeCastable` instead of `ValueCastable`. This ensures that `ValueCastable`s without a custom shape automatically get plain formatting.
+
+The default behavior proposed in this RFC ensures that a formatting implementation is always available, allowing generic code to print arbitrary values without worrying about an exception. Eg. something like `lib.fifo` could use debug prints when an element is added, automatically using rich formatting for shapes with `format`, while falling back to plain integers when rich formatting is not available.
+
+alternative default behaviors possible are:
+
+- raise `TypeError` (disallow formatting value-castables without explicitly implemented formatting)
+- no default, require every `ShapeCastable` to implement `format` (compatibility breaking)
+
+To avoid reserving a name and interfering with user code, `format` could be renamed to `_amaranth_format`.
```

Comment:
If it ever becomes a problem (no idea if it will) I think we should add something that works on all of the special methods; either "a `amaranth_` prefixed version takes priority" or "if `_amaranth_namespace_ = True` is defined on the class look for different methods".

However, I think it's not the place of this RFC to introduce this namespacing.

Reply:
Agreed.